### PR TITLE
RabbitMQ Structure & Consumers

### PR DESCRIPTION
**Pull request by binaryoverload**

### CODE REVIEW

File: src/main/kotlin/org/cascadebot/bot/rabbitmq/RabbitMQManager.kt [hunk 1]
```diff
@@ -23,6 +18,9 @@ class RabbitMQManager (config: RabbitMQ) {
     private val connectionFactory: ConnectionFactory = ConnectionFactory()
     private var connection: Connection
 
+    private var broadcastQueueName: String? = null
+        private set
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/bot/rabbitmq/RabbitMQManager.kt]: 1. why does this exist.
2. why is it not hardcoded

File: src/main/kotlin/org/cascadebot/bot/rabbitmq/consumers/BroadcastConsumer.kt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+                // TODO Filter for permissions when permission system is created
+                val mutualGuilds =
+                    Main.shardManager.getMutualGuilds(user, Main.shardManager.shards.first().selfUser).filter {
+                        val member = it.getMember(user) ?: return@filter false
+                        member.isOwner || member.hasPermission(Permission.ADMINISTRATOR)
+                    }
+
+                mutualGuilds.map { MutualGuildResponse(it) }
+            }
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/bot/rabbitmq/consumers/BroadcastConsumer.kt]: We really should be following the rabbitMQ data structure we laid out.
Talking about the request the bot is getting, not response. response is fine

File: src/main/kotlin/org/cascadebot/bot/rabbitmq/consumers/MetaConsumer.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                val responseObject = RabbitMQResponse.failure(
+                    StatusCode.BadRequest,
+                    ErrorCode.InvalidProperty,
+                    "The property '$property' is invalid"
+                )
+                channel.basicPublish("", properties.replyTo, replyProps, responseObject.toJsonByteArray())
+                channel.basicAck(envelope.deliveryTag, false)
+                return
+            }
+        }
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/bot/rabbitmq/consumers/MetaConsumer.kt]: Same as above

File: src/main/kotlin/org/cascadebot/bot/rabbitmq/objects/GeneralReqResObjects.kt [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        val InvalidJsonFormat = ErrorCode("invalid_json")
+
+        fun fromException(e: Exception): ErrorCode {
+            return ErrorCode(e.javaClass.simpleName)
+        }
+
+    }
+
+
+}
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/bot/rabbitmq/objects/GeneralReqResObjects.kt]: I have a feeling this object is going to get big. Maybe not have a giant enum for this.

File: src/main/kotlin/org/cascadebot/bot/rabbitmq/objects/GeneralReqResObjects.kt [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+        val InvalidJsonFormat = ErrorCode("invalid_json")
+
+        fun fromException(e: Exception): ErrorCode {
+            return ErrorCode(e.javaClass.simpleName)
+        }
+
+    }
+
+
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/bot/rabbitmq/objects/GeneralReqResObjects.kt]: Ahh yes double newline. Perfect formatting. /s

**Review by weeryan17 — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/cascadebot/bot/rabbitmq/objects/GeneralReqResObjects.kt (hunk 4 above)
Comment (weeryan17) [reply src/main/kotlin/org/cascadebot/bot/rabbitmq/objects/GeneralReqResObjects.kt]: We can do small objects split off by type of error if you want. So like with these codes we'd have an invalid category and a user category.

File: src/main/kotlin/org/cascadebot/bot/rabbitmq/RabbitMQManager.kt (hunk 1 above)
Comment (binaryoverload) [reply src/main/kotlin/org/cascadebot/bot/rabbitmq/RabbitMQManager.kt]: It's a queue with an auto-generated name which is why it needs to be stored. The queue only exists for the duration of the bot nodes connection.

File: src/main/kotlin/org/cascadebot/bot/rabbitmq/objects/GeneralReqResObjects.kt (hunk 4 above)
Comment (binaryoverload) [reply src/main/kotlin/org/cascadebot/bot/rabbitmq/objects/GeneralReqResObjects.kt]: What do you suggest?

File: src/main/kotlin/org/cascadebot/bot/rabbitmq/RabbitMQManager.kt (hunk 1 above)
Comment (weeryan17) [reply src/main/kotlin/org/cascadebot/bot/rabbitmq/RabbitMQManager.kt]: why does it need to be stored? Didn't look like it was being used anywhere, let me check again.

Comment (weeryan17) [reply src/main/kotlin/org/cascadebot/bot/rabbitmq/RabbitMQManager.kt]: Nvm I see it, although it's only needed during initialization, so probably should only be stored temporarily, and not on the class

Comment (binaryoverload) [reply src/main/kotlin/org/cascadebot/bot/rabbitmq/RabbitMQManager.kt]: It's used for the broadcast consumer, it maybe doesn't need to be stored. I'll have a think

File: src/main/kotlin/org/cascadebot/bot/rabbitmq/objects/GeneralReqResObjects.kt (hunk 4 above)
Comment (weeryan17) [reply src/main/kotlin/org/cascadebot/bot/rabbitmq/objects/GeneralReqResObjects.kt]: Have smaller objects that extends from error code that are separated out by category. So like in the above example we could have (obviously not exact, just an example)
```
class InvalidErrorCodes(): ErrorCode() {
    companion object {
        val InvalidProperty = InvalidErrorCodes("invalid_property")
        val InvalidMethod = InvalidErrorCodes("invalid_method")

        val InvalidJsonFormat = InvalidErrorCodes("invalid_json")
    }
}
```
```
class UserErrorCodes(): ErrorCode() {
    companion object {
        val UserNotFound = ErrorCode("user_not_found")
    }
}
```
Something like this, not exact obv, but just going through the list real quick (so not a complete list) of user errors for just custom commands we'd probably have have.
- User not found
- User not in guild
- Cannot set nick
- Cannot add role
- Cannot remove role

So i think keeping everything in one file is not practical considering that's just one object (and user is probably our smallest)

**Review by mrjvs — DISMISSED** [review]

**Review by weeryan17 — APPROVED** [review]

**Review by mrjvs — APPROVED** [review]

**Pull request merged**